### PR TITLE
Fix TransferAsset strange behavior

### PR DESCRIPTION
**Pull request by l4l**
### Description of the Change

Transfering asset with non-zero precision causes writing to WSV amount without precision and that basically lead to the strange things.

Particularly amount at [this line](https://github.com/hyperledger/iroha/pull/1470/files#diff-61f63c72eb4f7dc6082d4c9a2a4f8b5cL451) isn't properly set because it's not adjusted to the proper precision. Basically interested line is [this](https://github.com/hyperledger/iroha/pull/1470/files#diff-61f63c72eb4f7dc6082d4c9a2a4f8b5cL425) but it contains result, so there's no easy fix, so that's why the PR contains refactorings

### Benefits

Plus one test and minus one bug.
Also code for execution of TransferAsset was flattened (thus i believe simpler to read)

### Possible Drawbacks 

None

### Usage Examples or Tests *[optional]*

`make transfer_asset_test && test_bin/transfer_asset_test`

### CODE REVIEW

File: irohad/execution/impl/command_executor.cpp [hunk 1]
```diff
@@ -436,54 +436,54 @@ namespace iroha {
...
+          expected::Value<std::shared_ptr<shared_model::interface::Amount>>>(
+          amount_builder_.precision(asset.value()->precision())
+              .intValue(0)
+              .build());
+      return ((dest_account_asset | [](const auto &ast)
+                   -> boost::optional<const shared_model::interface::Amount &> {
+                return {ast->balance()};
+              })
+                  .get_value_or(*kZero.value)
+              + command.amount())
```
Comment (nickaleks) [inline irohad/execution/impl/command_executor.cpp]: Maybe use amount?

File: irohad/execution/impl/command_executor.cpp [hunk 2]
```diff
@@ -436,54 +436,54 @@ namespace iroha {
...
-              "account asset builder failed. reason " + *error.error,
-              command_name);
-        });
+    auto dest_account_asset_new = command_amount | [&](const auto &amount) {
+      const auto kZero = boost::get<
+          expected::Value<std::shared_ptr<shared_model::interface::Amount>>>(
+          amount_builder_.precision(asset.value()->precision())
+              .intValue(0)
+              .build());
+      return ((dest_account_asset | [](const auto &ast)
```
Comment (nickaleks) [inline irohad/execution/impl/command_executor.cpp]: Please move that to separate statement

File: irohad/execution/impl/command_executor.cpp (hunk 1 above)
Comment (nickaleks) [reply irohad/execution/impl/command_executor.cpp]: Also this behavior (implicit precision conversion) may have unexpected results, if we leave that we need to verify that it works as expected.

Comment (victordrobny) [reply irohad/execution/impl/command_executor.cpp]: As far as I understood it is necessary to have an amount with a precision of asset here

File: irohad/execution/impl/command_executor.cpp [hunk 3]
```diff
@@ -436,54 +436,55 @@ namespace iroha {
...
+          + *amount;
+      return new_amount | [this, &command](const auto &new_dest_balance) {
+        return account_asset_builder_.assetId(command.assetId())
+            .accountId(command.destAccountId())
+            .balance(*new_dest_balance)
+            .build();
+      };
+    };
+
+    auto map_error = [&command_name](auto t) {
```
Comment (nickaleks) [inline irohad/execution/impl/command_executor.cpp]: Please prefer pass by reference

File: irohad/execution/impl/command_executor.cpp [hunk 4]
```diff
@@ -436,54 +436,55 @@ namespace iroha {
...
+                        -> ExecutionResult {
+               return makeExecutionResult(
+                   commands->upsertAccountAsset(*src_amount) |
+                       [&] {
+                         return commands->upsertAccountAsset(*dst_amount);
+                       },
+                   command_name);
+             };
+           })
+        .match([](auto) -> ExecutionResult { return {}; },
```
Comment (nickaleks) [inline irohad/execution/impl/command_executor.cpp]: Why do you need match here?

File: test/integration/acceptance/transfer_asset_test.cpp [hunk 5]
```diff
@@ -445,3 +446,92 @@ TEST_F(TransferAsset, InterDomain) {
...
+                               IntegrationTestFramework::kDefaultDomain,
+                               precision)
+                  .build()
+                  .signAndAddSignature(kAdminKeypair)
+                  .finish())
+      .skipProposal()
+      .checkBlock([](auto &block) {
+        ASSERT_EQ(block->transactions().size(), 1) << "Cannot create asset";
+      })
+      .sendTx(proto::TransactionBuilder()
```
Comment (nickaleks) [inline test/integration/acceptance/transfer_asset_test.cpp]: A lot of duplication in transaction creation, maybe put it in a lambda

Comment (l4l) [reply test/integration/acceptance/transfer_asset_test.cpp]: should be done in IR-1463

**Pull request merged**